src/cell.cpp

```diff
@@ -107,13 +107,105 @@ vector<int32_t> tokenize(const std::string region_spec)
   return tokens;
 }
 
+//==============================================================================
+//! Add precedence for infix regions so intersections have higher
+//! precedence than unions using parenthesis.
+//==============================================================================
+
+std::vector<int32_t>::iterator add_parenthesis(
+  std::vector<int32_t>::iterator start, std::vector<int32_t>& infix)
+{
+  int32_t start_token = *start;
+  // Add left parenthesis
+  if (start_token == OP_INTERSECTION) {
+    start = infix.insert(start - 1, OP_LEFT_PAREN);
+  } else {
+    start = infix.insert(start + 1, OP_LEFT_PAREN);
+  }
+  start++;
+
+  // Initialize return iterator
+  std::vector<int32_t>::iterator return_iterator = infix.begin();
+
+  // Add right parenthesis
+  // While the start iterator is within the bounds of infix
+  while (start < infix.end()) {
+    start++;
+
+    // If the current token is an operator and is different than the start token
+    if (*start >= OP_UNION && *start != start_token) {
+      // Skip wraped regions but save iterator position to check precedence and
+      // add right parenthesis depending on the precedence of the original token
+      // when a right parenthesis is encountered of the opposite token
+      if (*start == OP_LEFT_PAREN) {
+        return_iterator = start;
+        int depth = 1;
+        do {
+          start++;
+          if (*start > OP_COMPLEMENT) {
+            if (*start == OP_RIGHT_PAREN) {
+              depth--;
+            } else {
+              depth++;
+            }
+          }
+        } while (depth > 0);
+      } else if (start_token == OP_UNION) {
+        start = infix.insert(start - 1, OP_RIGHT_PAREN);
+        if (return_iterator == infix.begin()) {
+          return_iterator = start - 1;
+        }
+        return return_iterator;
+      } else {
+        start = infix.insert(start, OP_RIGHT_PAREN);
+        if (return_iterator == infix.begin()) {
+          return_iterator = start - 1;
+        }
+        return return_iterator;
+      }
```

Comment:
These could be combined to simplify a bit:
```suggestion
      } else {
        start = infix.insert(
          start_token == OP_UNION ? start - 1 : start, OP_RIGHT_PAREN);
        if (return_iterator == infix.begin()) {
          return_iterator = start - 1;
        }
        return return_iterator;
      }
```